adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m

```diff
@@ -0,0 +1,199 @@
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "GADPangleRTBNativeRenderer.h"
+#import "GADMediationAdapterPangleConstants.h"
+#import <BUAdSDK/BUAdSDK.h>
+#import "GADMAdapterPangleUtils.h"
+#include <stdatomic.h>
+#import "GADPangleNetworkExtras.h"
+
+static NSString *const BUDNativeAdTranslateKey = @"bu_nativeAd";
+
+@interface GADPangleRTBNativeRenderer()<BUNativeAdDelegate> {
+    /// The completion handler to call when the ad loading succeeds or fails.
+    GADMediationNativeLoadCompletionHandler _loadCompletionHandler;
+    /// The Pangle native ad.
+    BUNativeAd *_nativeAd;
+    /// The Pangle related view
+    BUNativeAdRelatedView *_relatedView;
+    /// An ad event delegate to invoke when ad rendering events occur.
+    id<GADMediationNativeAdEventDelegate> _delegate;
+
+}
+
+@end
+
+@implementation GADPangleRTBNativeRenderer
+@synthesize images = _images,icon = _icon;
+
+- (void)renderNativeAdForAdConfiguration:(nonnull GADMediationNativeAdConfiguration *)adConfiguration
+                       completionHandler:(nonnull GADMediationNativeLoadCompletionHandler)completionHandler {
+    __block atomic_flag completionHandlerCalled = ATOMIC_FLAG_INIT;
+    __block  GADMediationNativeLoadCompletionHandler originalCompletionHandler = [completionHandler copy];
+    _loadCompletionHandler = ^id<GADMediationNativeAdEventDelegate> (_Nullable id<GADMediationNativeAd> ad, NSError *_Nullable error) {
+        if (atomic_flag_test_and_set(&completionHandlerCalled)) {
+            return nil;
+        }
+        id<GADMediationNativeAdEventDelegate> delegate = nil;
+        if (originalCompletionHandler) {
+            delegate = originalCompletionHandler(ad, error);
+        }
+        originalCompletionHandler = nil;
+        return delegate;
+    };
+    
+    NSString *placementId = adConfiguration.credentials.settings[GADMAdapterPanglePlacementID] ?: @"";
+    if (!placementId.length) {
+        NSError *error = GADMAdapterPangleErrorWithCodeAndDescription(GADPangleErrorInvalidServerParameters,
+                                                                      [NSString stringWithFormat:@"%@ cannot be nil.",GADMAdapterPanglePlacementID]
+                                                                      );
+        _loadCompletionHandler(nil, error);
+        return;
+    }
+    
+    GADPangleNetworkExtras *extras = adConfiguration.extras;
+    
+    BUAdSlot *slot = [[BUAdSlot alloc] init];
+    slot.ID = placementId;
+    slot.AdType = BUAdSlotAdTypeFeed;
+    slot.userData = extras.userDataString;
+    _nativeAd = [[BUNativeAd alloc]initWithSlot:slot];
```

Comment:
If the init fails, would there be a callback?

Reply:
I will add callback